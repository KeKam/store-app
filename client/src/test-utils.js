import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
