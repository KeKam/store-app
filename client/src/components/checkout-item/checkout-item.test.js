import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { render, screen } from '../../test-utils';
import {
  removeFromCart,
  addToCart,
  removeItem,
} from '../../redux/cart/cart.actions';
import CheckoutItem from './checkout-item';

describe('CheckoutItem component', () => {
  const mockState = {
    cartItems: [
      {
        id: 1,
        name: 'Winter falls',
        imageUrl: 'www.testImage.com',
        price: 100,
        quantity: 3,
      },
    ],
  };
  const mockCartItem = mockState.cartItems[0];

  it('should render CheckoutItem component', () => {
    const { asFragment } = render(<CheckoutItem cartItem={mockCartItem} />);

    expect(asFragment()).toMatchSnapshot();
    expect(
      screen.getByRole('img', { src: 'www.testImage.com' })
    ).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Winter falls');
    expect(screen.getByTestId('left-arrow')).toHaveTextContent('❮');
    expect(screen.getByTestId('quantity')).toHaveTextContent('3');
    expect(screen.getByTestId('right-arrow')).toHaveTextContent('❯');
    expect(screen.getByTestId('price')).toHaveTextContent('300 €');
    expect(screen.getByTestId('remove-button')).toHaveTextContent('✕');
  });

  it('should dispatch removeFromCart action on left arrow click', () => {
    const { getState, getDispatchedActions, rerenderWithRedux } = render(
      <CheckoutItem cartItem={mockCartItem} />,
      {
        initialState: { cart: mockState },
      }
    );

    userEvent.click(screen.getByTestId('left-arrow'));
    rerenderWithRedux(<CheckoutItem cartItem={getState().cart.cartItems[0]} />);

    expect(getDispatchedActions()).toContainEqual(removeFromCart(mockCartItem));
    expect(screen.getByText('2')).toBeInTheDocument();
  });

  it('should dispatch addToCart action on right arrow click', () => {
    const { getState, getDispatchedActions, rerenderWithRedux } = render(
      <CheckoutItem cartItem={mockCartItem} />,
      {
        initialState: { cart: mockState },
      }
    );

    userEvent.click(screen.getByTestId('right-arrow'));
    rerenderWithRedux(<CheckoutItem cartItem={getState().cart.cartItems[0]} />);

    expect(getDispatchedActions()).toContainEqual(addToCart(mockCartItem));
    expect(screen.getByText('4')).toBeInTheDocument();
  });

  it('should dispatch removeItem action on click', () => {
    const { getDispatchedActions } = render(
      <CheckoutItem cartItem={mockCartItem} />,
      {
        initialState: { cart: mockState },
      }
    );

    userEvent.click(screen.getByTestId('remove-button'));

    expect(getDispatchedActions()).toContainEqual(removeItem(mockCartItem));
  });
});
