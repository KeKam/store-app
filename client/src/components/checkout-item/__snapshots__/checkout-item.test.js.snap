// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem component should render CheckoutItem component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa kqBKPi"
  >
    <div
      class="sc-bwzfXH cBObtq"
    >
      <img
        alt="item"
        class="sc-htpNat flXzWO"
        src="www.testImage.com"
      />
    </div>
    <span
      class="sc-bxivhb bZgbkN"
      data-testid="title"
    >
      Winter falls
    </span>
    <span
      class="sc-bxivhb sc-ifAKCX ixeIqU"
    >
      <div
        class="sc-EHOje boCYJR"
        data-testid="left-arrow"
      >
        ❮
      </div>
      <span
        data-testid="quantity"
      >
        3
      </span>
      <div
        class="sc-EHOje boCYJR"
        data-testid="right-arrow"
      >
        ❯
      </div>
    </span>
    <span
      class="sc-bxivhb bZgbkN"
      data-testid="price"
    >
      300 €
    </span>
    <div
      class="sc-bZQynM cZxgzJ"
      data-testid="remove-button"
    >
      ✕
    </div>
  </div>
</DocumentFragment>
`;
